tokenizers/src/tokenizer/mod.rs

```diff
@@ -876,3 +900,35 @@ impl Tokenizer {
         }
     }
 }
+
+#[derive(Debug)]
+pub struct Lines<B> {
+    buf: B,
+}
+impl<B: BufRead> Iterator for Lines<B> {
+    type Item = std::io::Result<String>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut buf = String::new();
+        match self.buf.read_line(&mut buf) {
+            Ok(0) => None,
+            Ok(_n) => {
+                // if buf.ends_with('\n') {
+                //     buf.pop();
+                //     if buf.ends_with('\r') {
+                //         buf.pop();
+                //     }
+                // }
+                Some(Ok(buf))
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+fn lines_with_ending<B>(buf: B) -> Lines<B>
```

Comment:
To make this feel more "natural" (like the `.lines()`), we can maybe have a `trait` that requires a method `lines_with_ending` that does this, and we can implement it for `B: BufRead` so that we can then do `file.lines_with_ending().par_bridge()` on the `BufReader` above. What do you think?

Reply:
Much better idea, I'm new to rust and I think you can't implement trait for external structs, But it didn't knwo you could implement a trait for other traits. That's neat.

Reply:
You can actually implement any trait on any type, as long as one of them is defined in your own crate. It’s not possible if both are from external crates. This is known as the orphan rule, and it prevents having multiple implementations, with rust not knowing which one to use.

So you can either:
 - implement your own trait on any type
 - implement an external trait on any of your types
 - redefine your own type to just wrap another type, so that you can implement a trait on it.